src/_cffi_src/openssl/src/osrandom_engine.c

```diff
@@ -106,25 +155,25 @@ static int dev_urandom_fd(void) {
         }
     }
     if (urandom_cache.fd < 0) {
+#ifdef __linux__
+        if (wait_on_devrandom() < 0) {
+            goto error;
+        }
+#endif
+
         fd = open("/dev/urandom", O_RDONLY);
         if (fd < 0) {
             goto error;
         }
-        if (fstat(fd, &st)) {
+        if (set_cloexec(fd) < 0) {
             goto error;
         }
-        /* set CLOEXEC flag */
-        flags = fcntl(fd, F_GETFD);
-        if (flags == -1) {
-            goto error;
-        } else if (fcntl(fd, F_SETFD, flags | FD_CLOEXEC) == -1) {
+        if (fstat(fd, &st)) {
             goto error;
         }
         /* Another thread initialized the fd */
         if (urandom_cache.fd >= 0) {
```

Comment:
You're probably aware and don't care, but this double-checked idiom is racy: http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Reply:
this was the pre-existing race condition I mentioned, I'm not dealing with it in this PR :-)